docs/token-contracts/fa12/4-fa12-archetype.md

```diff
@@ -0,0 +1,447 @@
+```
+$ archetype fa12.arl > fa12.tz
+```
+
+> Install [Archetype](https://docs.archetype-lang.org/getting-started-1#installation) (minimum required version: 1.2.0).
+
+## Originate contract
+
+This section presents how to originate (deploy) the contract on the Tezos testnet with the Tezos client. It assumes you already have an activated account (if not follow this procedure).
+
+The following command generates the Micheline syntax for the initial storage:
+
+```
+$ archetype --set-caller-init tz1dEezCSqMVsmfnaAZS2rBJZfdFCZEyA7DP -t michelson-storage fa12.arl
+(Pair {  } { Elt "tz1dEezCSqMVsmfnaAZS2rBJZfdFCZEyA7DP" 10000000 })
+```
+
+The address passed as the initial caller (here `tz1dEezCSqMVsmfnaAZS2rBJZfdFCZEyA7DP`) is the address that possesses the total number of tokens (here 100000000). The generated initial storage is used to generate the origination command.
+
+The following command originates the contract:
+```
+$ tezos-client -S -A testnet-tezos.giganode.io -P 443 originate contract fa12 transferring 0 from alice running fa12.tz --init '(Pair {  } { Elt "tz1dEezCSqMVsmfnaAZS2rBJZfdFCZEyA7DP" 10000000 })' --burn-cap 0 --force -D
+```
+
+The cost of origination (at the time of writing) is **2.5ꜩ** on the testnet.
+
+It uses the [Giganode](https://tezos.giganode.io/) Tezos nodes infrastructure.
+
+> Install [Tezos client](../../../docs/setup/1-tezos-client).
+
+## Formal properties
+This section presents the properties verified by the originated contract.
+
+The contract declares two [assets](https://docs.archetype-lang.org/archetype-language/data-model):
+
+```archetype
+asset ledger identified by holder to big_map {
+  holder     : address;
+  tokens     : nat = 0;
+} initialized by {
+  { holder = caller; tokens = totalsupply }
+}
+```
+The `ledger` asset is the cap table: it holds the number of tokens for each token holder.
+`totalsupply` is the initial number of tokens held by the originator of the contract.
+
+```archetype
+asset allowance identified by addr_owner addr_spender to big_map {
+  addr_owner       : address;
+  addr_spender     : address;
+  amount           : nat;
+}
+```
+The `allowance` asset stores the amount of tokens that can be spent by `addr_spender` on the behalf of `addr_owner`.
+
+### Contract invariant
+
+```archetype
+ledger.sum(tokens) = totalsupply
+```
+> The total number of tokens is equal to `totalsupply` regardless of the state of the contract storage.
+
+### transfer
+
+```archetype
+entry %transfer (%from : address, %to : address, value : nat) {
+  require {
+    r1 otherwise "NotEnoughBalance" : ledger[%from].tokens >= value;
+  }
+  effect {
+    if caller <> %from then (
+      var current = allowance[(%from, caller)].amount;
+      dofailif(current < value, ("NotEnoughAllowance", ((value, current))));
+      allowance.update((%from, caller), { amount -=  value });
+    );
+    ledger.update(%from, { tokens -= value });
+    ledger.addupdate(%to, { tokens += value });
+  }
+}
+```
```

Comment:
A little prose here would be useful to follow along, something like ”This is the implementation of transfer `<insert-code>`...  , let's now see the various properties we can specify about its behavior" ...

Reply:
Done.